src/main/java/com/commercetools/paymenttoorderprocessor/jobs/actions/MessageReader.java

```diff
@@ -43,51 +44,72 @@
     @Value("${ctp.messagereader.minutesoverlapping}")
     private Integer minutesOverlapping;
 
-    private List<PaymentTransactionStateChangedMessage> messages = Collections.emptyList();
+    @Nonnull
+    private Queue<PaymentTransactionStateChangedMessage> unprocessedMessagesQueue = new ArrayDeque<>();
+
     private boolean wasInitialQueried = false;
     private long total;
     private long offset = 0L;
-    private final int RESULTSPERPAGE = 500;
-    private final int PAGEOVERLAP = 5;
-
-    private MessageQuery messageQuery;
+    final int RESULTSPERPAGE = 500;
+    final int PAGEOVERLAP = 5;
 
+    /**
+     * @return the oldest unprocessed message from the queue, or <b>null</b> if no new messages to process in CTP.
+     */
     @Override
+    @Nullable
     public PaymentTransactionStateChangedMessage read() {
         LOG.debug("wasInitialQueried: {}", wasInitialQueried);
-        if (isQueryNeeded()) {
-            getUnprocessedMessagesFromPlatform();
+
+        while (isQueryNeeded()) {
+            fetchUnprocessedMessagesFromPlatform();
         }
-        return getMessageFromList();
+
+        return getUnprocessedMessageFromQueue();
     }
 
-    private PaymentTransactionStateChangedMessage getMessageFromList() {
-        if (messages.isEmpty()) {
-            return null;
-        } else {
-            timeStampManager.setActualProcessedMessageTimeStamp(messages.get(0).getLastModifiedAt());
-            return messages.remove(0);
-        }
+    /**
+     * @return oldest unprocessed message from the queue if exists, or <b>null</b>
+     */
+    @Nullable
+    private PaymentTransactionStateChangedMessage getUnprocessedMessageFromQueue() {
+        return unprocessedMessagesQueue.poll();
     }
 
+    /**
+     * A first or next CTP query is needed if we don't have unprocessed messages in the queue.
+     *
+     * @return <b>true</b> if messages queue never fetched or {@code unprocessedMessagesQueue} is empty and CTP still
+     * has items to fetch (query next page).
+     */
     private boolean isQueryNeeded() {
-        return !wasInitialQueried || (messages.isEmpty() && total > offset);
+        return !wasInitialQueried || (unprocessedMessagesQueue.isEmpty() && total > offset);
```

Comment:
So you don't fetch all unprocessed messages, but only until you get some unprocessed messages? What's the reason for this?

Reply:
because there might be hundred thousands of them, but we process them one by one on chronicle order (to avoid inconsistency, as you mentioned in the [comment above](https://github.com/commercetools/commercetools-payment-to-order-processor/pull/27#discussion_r180036794))

Reply:
this misunderstanding in here:
> And in the next cronjob run, it will fetch the next 500

  - in the first (current) run the job stores `lastProcessedMessageTimeStamp = message[500].lastModifiedAt`
  - in the next cron job run it will fetch the first 500 messages since `lastProcessedMessageTimeStamp`